screeps-game-api/src/macros.rs

```diff
@@ -375,3 +375,182 @@ macro_rules! game_map_access {
         )*
     };
 }
+
+/// Get a value from memory given a path, returning `None` if any thing along the way does not
+/// exist.
+///
+/// # Examples
+///
+/// Get a reference with type u32 at the path creeps.John.count.
+///
+/// ```no_run
+/// #[macro_use]
+/// extern crate screeps;
+///
+/// # fn main() {
+/// let mem = screeps::memory::root();
+/// let val = mem_get!(mem.creeps.John.count.int);
+/// # }
+/// ```
+///
+/// Get something using a variable path.
+///
+/// ```no_run
+/// #[macro_use]
+/// extern crate screeps;
+///
+/// # fn main() {
+/// let mem = screeps::memory::root();
+/// let creep_name = "John";
+/// let what_to_get = "count";
+/// let val1 = mem_get!(mem.creeps[creep_name][what_to_get].int);
+/// let val2 = mem_get!(mem.creeps[creep_name].count.int);
+/// assert_eq!(val1, val2);
+/// # }
+/// ```
+///
+/// Accepted suffixes for type are methods that exist on `MemoryReference`, such as `num`, `int`,
+/// `string`, `bool`, `arr` and `dict`.
+#[macro_export]
+macro_rules! mem_get {
+    // Macro entry point
+    ($memory_reference:ident $($rest:tt)*) => {
+        mem_get!(@so_far { Some(&$memory_reference) } @rest $($rest)*)
+    };
+    // Access the last part with a variable
+    (@so_far { $reference_so_far:expr } @rest [ $next_part_variable:expr ] . $accessor:ident) => {
+        $reference_so_far.and_then(|v| v.$accessor($next_part_variable))
+    };
+    // Access the last part with a hardcoded ident
+    (@so_far { $reference_so_far:expr } @rest . $next_part:ident . $accessor:ident) => {
+        $reference_so_far.and_then(|v| v.$accessor(stringify!($next_part)))
+    };
+    // Access the next (but not last) part with a variable
+    (@so_far { $reference_so_far:expr } @rest [ $next_part_variable:expr ] $($rest:tt)+) => {
+        mem_get!(
+            @so_far {
+                $reference_so_far.and_then(|v| v.dict($next_part_variable))
+            }
+            @rest $($rest)*
+        )
+    };
+    // Access the next (but not last) part with a hardcoded ident
+    (@so_far { $reference_so_far:expr } @rest . $next_part:ident $($rest:tt)+) => {
+        mem_get!(
+            @so_far {
+                $reference_so_far.and_then(|v| v.dict(stringify!($next_part)))
+            }
+            @rest $($rest)*
+        )
+    };
+    ($($not_valid:tt)*) => {
+        compile_error!(concat!("Unexpected usage of mem_get! usage: ", stringify!($($not_valid)*)))
+    }
+}
+
+/// Set a value in memory given a path, creating dicts for intermediate places if they do not exist.
+///
+/// # Panics
+///
+/// This macro will panic if any of the intermediate paths exist but are not dictionaries.
```

Comment:
Isn't that harsh? Maybe return an `Error` instead?

Reply:
I'd consider a path which should be a dictionary being an array or integer a hard error mostly because it should never happen outside of logic errors, and there isn't one generally good way to recover from it without the potential of loosing data.

If two parts of code are accessing the same memory path and expecting it to be different types, that's a bug. I mean we could return an Option or Err here, but I would expect that most usages of this utility wouldn't want to have to think about handing this situation besides just panicking themselves.

Reply:
I for one would rather it fails without panicking and write an `error!` to the logs.

Your argument basically boils down to: in order to not have corrupt memory, I'll stop executing my AI midway in the game loop. To me the latter is much worse, especially since I'm not given a choice (afaik, you can't catch a `panic`).

Reply:
I mean, it *is* a bug, and should be fixed, but _in prod_, I'd rather a bit of lost memory than a non-functional colony.

Reply:
True! I'll switch this to returning a `Result`.